### ISSUE TITLE
Fullscreen keyboard toggle

### ISSUE THREAD

**Issue by tuncer** [ArturKovacs/emulsion#30]
I wanted to implement fullscreen toggle via keyboard and saw that it is ~~semi-implemented~~ already. Thus, before I make the wrong diff, let's discuss how you'd like fullscreen toggle to work. Should the binding be configurable? What about key handling in picture_widget and picture_panel, meaning, where should it be implemented?

**Comment by ArturKovacs** [comment]
Currently doub-clicking with the left mouse button toogles fullscreen mode so I'm not sure what you mean by semi-implemented. Although making the binding configurable as you suggested would be a welcome addittion.

Oh no! I forgot to remove `picture_panel` after the ui rewrite. That module is completely obsolete and was replaced by `picture_widget`.

With that said most if not all input handling is done inside the `picture_widget` may not be very clean, the input handling could be in a separate module instead calling higher level functions from the picture widget like `next_image()` and `fit_img_to_window()` and such.

To get back to bindings for toggling full-screen, it's a bit different than regular keyboard input becaouse most importantly it's a double click, or with more general terms two of the same type of input happening in short succession. I think the biggest challange is to find a good way to represent this in the bindings section of the configuration. As an initial proposal I think the following could be a how this is expressed
```toml
# two times the left mouse button
toggle_fullscreen = "double:lmb"
# two times alt+f
toggle_fullscreen = "double:alt+f"
# two times ':' but this probably won't happen (space can be ommited)
toggle_fullscreen = "double: :"
```

**Comment by tuncer** [comment]
You're right, and I didn't think of looking for a mouse binding. Fixed the description.

**Comment by tuncer** [comment]
If I understand you correctly, you're about to cleanup/refactor, thereby removing `picture_panel`, right?

**Comment by tuncer** [comment]
Can't we toggle fullscreen with a single `Alt+F`?

**Comment by tuncer** [comment]
I'd implement `Alt+F` and `F11` because some laptops make you press the `Fn` key to access F keys.

**Comment by ArturKovacs** [comment]
>f I understand you correctly, you're about to cleanup/refactor, thereby removing picture_panel, right?

The refactoring was done at the v1.9 release. I just forgot to remove the `picture_panel.rs` file.

> Can't we toggle fullscreen with a single Alt+F (and F11 for discoverability)?

The bindings I listed above were just examples of the syntax I proposed. Of course if the users would omit `double:` the binding would mean that a single input toggles fullscreen. I updated my previous comment to reflect this.

Once custom bindings are implemented for this, the users will be able to set this to whatever they please :) Although I like the idea to add F11 as a default for toogling fullscreen.

**Comment by tuncer** [comment]
I'd propose to include `Alt-F` in addition to `F11` as default bindings, since some keyboards require you to press `Fn` to trigger `F` keys.

**Comment by tuncer** [comment]
See #31.

**Comment by tuncer** [comment]
BTW, is it possible `cargo fmt` wasn't run after some of the changes? You can test with `cargo run --all -- --check`, which is one of the steps to include in a pre-merge GitHub Action.

**Comment by ArturKovacs** [comment]
> I'd propose to include Alt-F in addition to F11 as default bindings, since some keyboards require you to press Fn to trigger F keys.

I don't really like `alt+f` for full screen 😄 I know I wrote that but that was just a silly example. My feeling is that if someone doesn't want to double click or press `F11` / `Fn+F11` then they will probably be ok with setting their own binding in the config file. Plus `Fn+F11` is two keys just like `alt+f`. Another element behind my reasoning is that I can't imagine a use case when entering and leaving full-sceen mode is something that needs to happen a lot.

**Comment by ArturKovacs** [comment]
> 
> 
> BTW, is it possible `cargo fmt` wasn't run after some of the changes? You can test with `cargo run --all -- --check`, which is one of the steps to include in a pre-merge GitHub Action.

Yes I haven't run `cargo fmt` in a while. If you would add a Github Action that makes that check when requesting a pull into the `master`, I'd be happy to merge it.

**Comment by tuncer** [comment]
Done, see #32. I don't know if it builds on macOS, but it's included. Also, I don't think I can configure the workflow and think you have to do that as project owner.

To create releases, you add a second workflow which is only triggered on a new tag. In that workflow, you compile release binaries, package them up, and upload them to the github release. Actually, you can automate publishing the crates, too, again when a new tag is pushed.

**Comment by tuncer** [comment]
I agree with your Alt+F argument and, even though it's admittedly not a frequent action, maybe we can find a binding that is easy to reach with one hand and doesn't require pressing Fn on some keyboards. Either way, let's get keyboard toggle implemented first. We can figure out default bindings later, if needed.

**Comment by ArturKovacs** [comment]
>Either way, let's get keyboard toggle implemented first. We can figure out default bindings later, if needed.

Good thinking. Are you going to start working on this?

**Comment by tuncer** [comment]
Done: #36

**Comment by ArturKovacs** [comment]
I'm closing this as the implementation was merged by #36
